modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/classloader/AbstractArtifactClassLoaderConfigurationAssembler.java

```diff
@@ -0,0 +1,166 @@
+import static org.mule.runtime.module.artifact.api.descriptor.BundleDescriptor.MULE_PLUGIN_CLASSIFIER;
+
+import static java.lang.String.format;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import com.google.common.collect.ImmutableSet;
+import org.mule.runtime.api.exception.MuleRuntimeException;
+import org.mule.runtime.module.artifact.activation.api.ArtifactActivationException;
+import org.mule.runtime.module.artifact.api.descriptor.BundleDependency;
+import org.mule.runtime.module.artifact.api.descriptor.ClassLoaderModel;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Assembles the class loader configuration for an artifact given all its pieces.
+ */
+public abstract class AbstractArtifactClassLoaderConfigurationAssembler {
+
+  private static final Logger LOGGER = getLogger(AbstractArtifactClassLoaderConfigurationAssembler.class);
+
+  protected org.mule.tools.api.classloader.model.ClassLoaderModel packagerClassLoaderModel;
+
+  public AbstractArtifactClassLoaderConfigurationAssembler(org.mule.tools.api.classloader.model.ClassLoaderModel packagerClassLoaderModel) {
+    this.packagerClassLoaderModel = packagerClassLoaderModel;
+  }
+
+  public ClassLoaderModel createClassLoaderModel() {
+    ArtifactClassLoaderConfigurationBuilder artifactClassLoaderConfigurationBuilder =
+        new ArtifactClassLoaderConfigurationBuilder(packagerClassLoaderModel, getProjectFolder());
+
+    artifactClassLoaderConfigurationBuilder
+        .exportingPackages(getExportedPackages())
+        .exportingResources(getExportedResources());
+    // TODO: add privileged packages and resources
+
+    // TODO: analyze patched dependencies in classloader-model-patch.json
+
+    final List<URL> dependenciesArtifactsUrls =
+        loadUrls(getProjectFolder(), getBundleDependencies(), artifactClassLoaderConfigurationBuilder);
+    dependenciesArtifactsUrls.forEach(artifactClassLoaderConfigurationBuilder::containing);
+
+    // TODO: check if it belongs to the deny-list to decide whether local packages should be populated
+    populateLocalPackages(packagerClassLoaderModel, artifactClassLoaderConfigurationBuilder);
+
+    artifactClassLoaderConfigurationBuilder.dependingOn(new HashSet<>(getBundleDependencies()));
+
+    return artifactClassLoaderConfigurationBuilder.build();
+  }
+
+  private List<URL> loadUrls(File artifactFile,
+                             List<BundleDependency> dependencies,
+                             ArtifactClassLoaderConfigurationBuilder artifactClassLoaderConfigurationBuilder) {
+    final List<URL> dependenciesArtifactsUrls = new ArrayList<>();
+
+    // TODO: consider artifact patches for the case this is run within a Runtime
```

Comment:
yup

Reply:
Created user story [W-11203071](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000xoVUsYAM/view) to tackle it.